### PR TITLE
mount: add new LazyUnmount= setting for mount units, mapping to umoun…

### PR DESCRIPTION
**Pull request by brulon**
…t(8)'s "-l" switch

### CODE REVIEW

**Comment by arvidjaar** [comment]
Why is it needed?

**Comment by poettering** [comment]
Is this patch really sufficient? Note that systemd tracks active mounts via /proc/self/mountinfo, and expects that after /bin/umount returned the mount it was called for disappeared from /proc/self/mountinfo. Is that actually what happens if lazy mounts are used?

**Comment by brulon** [comment]
@arvidjaar 

> Why is it needed?

If a mount unit is created for a hot pluggable device (i.e., USB thumb drive), and the device is unplugged while an application is still accessing a file on the device, a standard umount will fail.  In this scenario, the mount point will remain active (still shows up in /proc/self/mountinfo) even though the drive is no longer connected.  Subsequent attempts at plugging in the thumb drive will not cause the drive to be re-mounted (presumably because systemd thinks the drive is still mounted).

A lazy umount takes care of this scenario.  From umount man page ... "Note that a file system cannot be unmounted when it is 'busy' - for example, when there are open files on it...A lazy unmount avoids this problem."

I suppose a force would also work, but the lazy option provides the proper filesystem cleanup

**Comment by poettering** [comment]
@brulon the usb removal thing really is a good example where a forced, not a lazy unmount is appropriate...

**Comment by brulon** [comment]
@poettering 

> Is this patch really sufficient?

I verified that with a lazy umount the device will be removed from /proc/self/mountinfo

**Comment by brulon** [comment]
@poettering , why is forced more appropriate than lazy in this scenario?

**Comment by brulon** [comment]
@arvidjaar @poettering 
Regardless of scenario, is there a reason why we wouldn't supply the user with the options available to mount/umount as long as the functionality is tested?  In the event that a umount -l fails, the outcome is really no different than if a straight umount fails.

I think it would be appropriate for me to go ahead and create a pull request for the force option if you feel that this option is warranted.

**Comment by poettering** [comment]
@brulon well, "lazy" umount is really for stuff where you say "keep the objects around as long as they are referenced, but don't allow opening of new objects". the "force" umount mode otoh is for stuff where you say "invalidate all objects, immediately". If a medium vanishes due to unplug, then the objects are simply invalid, as every write or read access will necessarily result in an IO error from that point on. Hence you really should use a force umount in this case, rather than a lazy one.

If the lazy umounts disappear immediately from /proc/self/mountinfo then I figure it's OK to merge this patch.

File: man/systemd.mount.xml [hunk 1]
```diff
@@ -352,6 +352,18 @@
       </varlistentry>
 
       <varlistentry>
+        <term><varname>LazyUnmount=</varname></term>
+
+        <listitem><para>Takes a boolean argument. If true, detach the
+        filesystem from the filesystem hierarchy now, and cleanup all
```
Comment (poettering) [inline man/systemd.mount.xml]: "now" → "at time of the unmount operation".

"cleanup" → "clean up".

File: man/systemd.mount.xml [hunk 2]
```diff
@@ -352,6 +352,18 @@
...
+        references to the filesystem as soon as it is not busy anymore.
```
Comment (poettering) [inline man/systemd.mount.xml]: as it is → as they are

**Comment by poettering** [comment]
so, the usecase is bad, but I think we should probably merge this anyway, as it is useful for other cases I think (consider: a tmpfs mount rather than a real medium, where it's completely OK to let programs continue use their open files on the fs as long as they want). Code looks good, man page needs a few more fixes, see above.

That all said, maybe if we cover lazy umounts we should also cover forced umounts at the same time. For example, we could add "UnmountMode=" which takes an enum: "lazy", "force" and "regular". (the latter is a bad name, not sure what a better one would be. Maybe "try"?

**Comment by brulon** [comment]
@poettering thanks for all the feedback.

Regarding moving to "UnmountMode="...I had thought about using the same approach, but I decided to follow how the mount flags were handled (i.e., each supported flag is a separate [MOUNT] option).
Options= (-o)
SloppyOptions= (-s)
Type= (-t)
I figured consistency was the best approach.  However, looking at this again, I realize that 2 of those mount flags take additional input (making it harder to collapse), and the "force" and "lazy" options share commonality in that they are methods of unmounting.

Would it be sufficient to just say leaving UnmountMode= out of the mount unit is equivalent to "regular" mode?  Essentially saying that using UnmountMode= implies that you will be setting a flag that changes the mode.

If you feel that keeping an enum for the default behavior is necessary, maybe we could use "default" or "standard" instead of "regular".

**Comment by poettering** [comment]
well, the other options either are true booleans, or have real parameters. But the lazy vs. forced vs. try thing is exclusive. If there'd be two options LazyUnmount= and ForceUnmount= and both would be true, what would that even mean? But if we'd expose this as enum instead, this problem would go away neatly.

And yes, the idea would be that UnmountMode=normal (or whatever we call it) would be the implied default, compatible with current systemd versions.

**Comment by brulon** [comment]
> But the lazy vs. forced vs. try thing is exclusive. If there'd be two options LazyUnmount= and 
> ForceUnmount= and both would be true, what would that even mean?

I'm not sure if I am following this.  Per umount(8) and umount(2), Force and Lazy are 2 completely separate boolean flags, and there is nothing keeping a user from selecting both flags at the same time.  Now, I don't know why a user would want to do this, but it's not forbidden by umount.

I think it's fine if you want to forbid the 2 flags from being simultaneously set in systemd, but it doesn't follow the way umount deals with the options.

**Comment by arvidjaar** [comment]
> the "force" umount mode otoh is for stuff where you say "invalidate all objects, immediately". If a medium vanishes due to unplug, then the objects are simply invalid, as every write or read access will necessarily result in an IO error from that point on. Hence you really should use a force umount in this case, rather than a lazy one.

Not really. "force" is misnomer - it calls fs-specific code with intention to inform filesystem that umount is pending. Only very small number of filesystems actually supports it and kernel does not even checks for result, it is completely up to fs to do anything useful in this case. In particular, none of filesystems usually used for local medium supports it, so this is entirely noop in most cases. And when I really needed it (NFS) it usually hung anyway :)

> If there'd be two options LazyUnmount= and ForceUnmount= and both would be true, what would that even mean?

Kernel is fine with it.  That would mean that kernel first calls fs-specific code and then tries to detach subtree. If there is nothing opened on this subtree presumably it will be unmounted normally as result.

What I do not really like in this proposal is that both options are for emergency only. We really do not want to do lazy umount _every_ time as part of normal operation. But we cannot predict emergency when we create unit definition. This is something that requires manual administrator intervention.

**Comment by brulon** [comment]
> What I do not really like in this proposal is that both options are for emergency only. We really do not want to do lazy umount every time as part of normal operation. But we cannot predict emergency when we create unit definition. This is something that requires manual administrator intervention. 

I didn't realize lazy umount was considered emergency-only, but I now see in some umount documentation that it was intended to be used if a standard umount fails "for no apparent reason". The same doc mentions finding and killing processes that cause standard umount to return "busy". 

My question is whether you would want systemd to be responsible for finding and killing processes in the "busy" case?

At the very least, could this behavior be changed to always try standard umount first, and only try the "emergency" options (like -l or -f) if the options are set and standard umount fails?

**Comment by poettering** [comment]
if the flags are indepdent, then we should expose them as two separate boolean props I figure. Would be happy to take a patch for that.

I am very much against the killing spree thing. It's slow, racy, and hacky to do such a thing. It would be OK to do this if the kernel would have frevoke() or something like that, but otherwise I think this should not be done by systemd.

**Comment by poettering** [comment]
@brulon if you make the requested changes to the man page the patch could go in right away, and we can add the other flag in a separate PR. Alternatively, you could update this PR to add both at the same time, it's up to you.

**Comment by brulon** [comment]
@poettering I think I would rather just make the requested changes to this PR so we can close this out.  I will follow up later with a PR for the "force" option.  Will update in the next day or so.

**Comment by brulon** [comment]
@poettering I think I would rather just make the requested changes to this PR so we can close this out.  I will follow up later with a PR for the "force" option.  Will update in the next day or so.

**Comment by brulon** [comment]
Requested changes to man page made, squashed into original commit and branch has been updated.  I also realized in the process that a necessary change had been omitted from the original pull request (to core/load-fragment-gperf.gperf.m4).  This change has also been added to the branch.

**Comment by poettering** [comment]
looks pretty OK. But please make two more changes: please print this in "mount_dump()", so that "systemd-analyze dump" shows this value.

Please also make sure this works with "systemd-mount -p", i.e. for transient mount units. For that make sure that bus_mount_set_transient_property() can parse this if it is passed over the bus (see dbus-mount.c), and that bus_append_unit_property_assignment() can parse it when passed on the cmdline (see bus-unit-util.c).

**Comment by brulon** [comment]
updated PR with requested changes to mount_dump().

still need to verify LazyUnmount works for transient mount units. Will test this soon and report back when tested.

**Comment by brulon** [comment]
@poettering 

> Please also make sure this works with "systemd-mount -p", i.e. for transient mount units.

Doesn't "-p" specifically deal with [Unit] properties (not [Mount] properties)?  LazyUnmount is a [Mount] property.  Are you wanting me to create a new option to systemd-mount -- like "-l --lazy-unmount"?

**Comment by poettering** [comment]
"-p" applies to all kinds of props of the unit, regardless if exposed on the [Unit] or [Mount] unit file section/bus interface.

**Comment by brulon** [comment]
[Mount] props appear to be poorly supported by the mount-tool.  I tested the following existing props...

systemd-mount /dev/sdb1 /media/barron/thumb --property=SloppyOptions=true
systemd-mount /dev/sdb1 /media/barron/thumb --property=DirectoryMode=0555
systemd-mount /dev/sdb1 /media/barron/thumb --property=TimeoutSec=1

All of these produced the following error...
**Unknown assignment <fill in entered property>.
Failed to create bus message: Invalid argument**

systemd-mount /dev/sdb1 /media/barron/thumb --property=Type=vfat
This one worked, but is already supported via --type option

Should we hold up this PR for this?  It would seem that there needs to be a separate effort to deal with general [Mount] prop support.

**Comment by poettering** [comment]
yes, many of the mount properties are not accessible yet for transient units, but that's mostly because the concept of transient units wasn't available for mounts until recently. For the service properties we were in a similar situation, but since that try to always expose the new settigns added for transient units as well, from day 1. We should really do the same for mount units too I figure. The code for that is actually not complicated...

But anyway, will merge this for now, but we really should get this fixed in a later patch.

**Comment by poettering** [comment]
This also fixed #1257 btw.

**Pull request merged**